File: app/src/main/java/com/fire/mangareader/data/download/DownloadManager.java
package com.fire.mangareader.data.download;

import android.content.Context;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.fire.mangareader.data.service.DownloadService;

import java.util.HashMap;
import java.util.Map;

public class DownloadManager {
    private final Context context;
    
    // Holds the progress of currently downloading chapters
    // Key: chapterUrl, Value: Integer progress (0-100)
    private final MutableLiveData<Map<String, Integer>> downloadProgress = new MutableLiveData<>(new HashMap<>());

    
    
    
    // ADDED FOR COMPOSE COMPATIBILITY
    public void enqueueDownload(String chapterId, String mangaId, String mangaTitle, String chapterTitle) {
        startDownload(mangaId, chapterId, chapterTitle);
    }

    public kotlinx.coroutines.flow.Flow<java.util.List<com.fire.mangareader.data.local.entity.DownloadedChapter>> getDownloadsFlow() {
        return kotlinx.coroutines.flow.FlowKt.emptyFlow();
    }
    
    public void deleteDownload(String chapterId) {
        cancelDownload(chapterId);
    }



    public DownloadManager(Context context) {
        this.context = context.getApplicationContext();
    }

    public void startDownload(String mangaUrl, String chapterUrl, String chapterTitle) {
        DownloadService.startDownload(context, mangaUrl, chapterUrl, chapterTitle);
    }

    public void cancelDownload(String chapterUrl) {
        // Find and cancel
        // For now, DownloadService only supports cancelAll
        DownloadService.cancelAllDownloads(context);
    }
    
    public void cancelAllDownloads() {
        DownloadService.cancelAllDownloads(context);
    }

    public void updateProgress(String chapterUrl, int progress) {
        Map<String, Integer> current = downloadProgress.getValue();
        if (current == null) current = new HashMap<>();
        current.put(chapterUrl, progress);
        downloadProgress.postValue(current);
    }
    
    public void removeProgress(String chapterUrl) {
        Map<String, Integer> current = downloadProgress.getValue();
        if (current != null && current.containsKey(chapterUrl)) {
            current.remove(chapterUrl);
            downloadProgress.postValue(current);
        }
    }

    public LiveData<Map<String, Integer>> getDownloadProgress() {
        return downloadProgress;
    }
}

File: app/src/main/java/com/fire/mangareader/data/download/DownloadManager.kt
package com.fire.mangareader.data.download

import android.content.Context
import androidx.work.*
import com.fire.mangareader.data.local.AppDatabase
import com.fire.mangareader.data.local.entity.DownloadedChapter
import kotlinx.coroutines.flow.Flow

class DownloadManager(private val context: Context) {

    private val db = AppDatabase.getDatabase(context)
    private val downloadDao = db.downloadDao()
    private val workManager = WorkManager.getInstance(context)

    fun getDownloadsFlow(): Flow<List<DownloadedChapter>> {
        return downloadDao.getAllDownloads()
    }

    suspend fun enqueueDownload(
        chapterId: String,
        mangaId: String,
        mangaTitle: String,
        chapterTitle: String
    ) {
        val existing = downloadDao.getDownload(chapterId)
        if (existing?.state == 2) {
            // Already downloaded
            return
        }

        // Insert as QUEUED
        val download = DownloadedChapter(
            chapterId = chapterId,
            mangaId = mangaId,
            mangaTitle = mangaTitle,
            chapterTitle = chapterTitle,
            totalPages = 0,
            downloadedPages = 0,
            state = 0, // QUEUED
            downloadPath = ""
        )
        downloadDao.insertDownload(download)

        val data = workDataOf(
            DownloadWorker.KEY_CHAPTER_ID to chapterId,
            DownloadWorker.KEY_MANGA_ID to mangaId,
            DownloadWorker.KEY_MANGA_TITLE to mangaTitle,
            DownloadWorker.KEY_CHAPTER_TITLE to chapterTitle
        )

        val request = OneTimeWorkRequestBuilder<DownloadWorker>()
            .setInputData(data)
            .setExpedited(OutOfQuotaPolicy.RUN_AS_NON_EXPEDITED_WORK_REQUEST)
            .build()

        workManager.enqueueUniqueWork(
            "download_$chapterId",
            ExistingWorkPolicy.KEEP,
            request
        )
    }

    suspend fun cancelDownload(chapterId: String) {
        workManager.cancelUniqueWork("download_$chapterId")
        val existing = downloadDao.getDownload(chapterId)
        if (existing != null) {
            downloadDao.updateDownload(existing.copy(state = 3)) // Error/Canceled
        }
    }

    suspend fun deleteDownload(chapterId: String) {
        cancelDownload(chapterId)
        val existing = downloadDao.getDownload(chapterId)
        if (existing != null) {
            val file = java.io.File(existing.downloadPath)
            if (file.exists()) {
                file.deleteRecursively()
            }
            downloadDao.deleteDownloadById(chapterId)
        }
    }
}
